controllers/storageclassclaim_controller.go

```diff
@@ -0,0 +1,484 @@
+		)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		storageClass := &storagev1.StorageClass{}
+		storageClass.Name = r.storageClassClaim.Name
+		if err = r.get(storageClass); err != nil && !errors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("failed to get StorageClass %s: %s", storageClass.Name, err)
+		}
+		if storageClass.UID != "" {
+
+			if err = r.delete(storageClass); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to delete StorageClass %s: %s", storageClass.Name, err)
+			}
+		} else {
+			r.log.Info("StorageClass already deleted.")
+		}
+		if contains(r.storageClassClaim.GetFinalizers(), StorageClassClaimFinalizer) {
+			r.storageClassClaim.Finalizers = remove(r.storageClassClaim.Finalizers, StorageClassClaimFinalizer)
+			if err := r.update(r.storageClassClaim); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from storageClassClaim: %s", err)
+			}
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *StorageClassClaimReconciler) getCephFSStorageClass(data map[string]string) *storagev1.StorageClass {
+	pvReclaimPolicy := corev1.PersistentVolumeReclaimDelete
+	allowVolumeExpansion := true
+	storageClass := &storagev1.StorageClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.storageClassClaim.Name,
+			Namespace: r.storageClassClaim.Namespace,
+			Annotations: map[string]string{
+				"description": "Provides RWO and RWX Filesystem volumes",
+			},
+		},
+		ReclaimPolicy:        &pvReclaimPolicy,
+		AllowVolumeExpansion: &allowVolumeExpansion,
+		Provisioner:          fmt.Sprintf("%s.cephfs.csi.ceph.com", r.ocsClient.Namespace),
+		Parameters:           data,
+	}
+	return storageClass
+}
+
+func (r *StorageClassClaimReconciler) getCephRBDStorageClass(data map[string]string) *storagev1.StorageClass {
+	pvReclaimPolicy := corev1.PersistentVolumeReclaimDelete
+	allowVolumeExpansion := true
+	storageClass := &storagev1.StorageClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.storageClassClaim.Name,
+			Namespace: r.storageClassClaim.Namespace,
+			Annotations: map[string]string{
+				"description": "Provides RWO Filesystem volumes, and RWO and RWX Block volumes",
+			},
+		},
+		ReclaimPolicy:        &pvReclaimPolicy,
+		AllowVolumeExpansion: &allowVolumeExpansion,
+		Provisioner:          fmt.Sprintf("%s.rbd.csi.ceph.com", r.ocsClient.Namespace),
+		Parameters:           data,
+	}
+	return storageClass
+}
+
+func (r *StorageClassClaimReconciler) createOrReplaceStorageClass(storageClass *storagev1.StorageClass) error {
+	existing := &storagev1.StorageClass{}
+	existing.Name = r.storageClassClaim.Name
+
+	if err := r.get(existing); err != nil && !errors.IsNotFound(err) {
+		return fmt.Errorf("failed to get StorageClass: %v", err)
+	}
+
+	// If present then compare the existing StorageClass with the received StorageClass, and only proceed if they differ.
+	if reflect.DeepEqual(existing.Parameters, storageClass.Parameters) {
+		return nil
+	}
```

Comment:
We should have moved this inside `existing.UID` check as it make sense to have comparison only when storageclass exists